File: dart_attempt/lib/src/move.dart
part of sixstair;

abstract class Move {
  static const FLIP = const Flip();
  
  factory Move.fromString(String s) {
    if (s == 'F') {
      return FLIP;
    }
    Match m = new RegExp(r'^T(-?[1-6])$').matchAsPrefix(s);
    if (m == null) throw new ArgumentError('invalid move: $s');
    return new Turn(int.parse(m[1]));
  }
  
  void apply(SixStair s);
}

class Turn implements Move {
  final int turns;
  
  const Turn(this.turns);
  
  void apply(SixStair s) {
    if (turns > 0) {
      for (int i = 0; i < turns; i++) {
        s.turn(1);
      }
    } else {
      for (int i = 0; i < -turns; i++) {
        s.turn(-1);
      }
    }
  }
  
  String toString() {
    return 'T$turns';
  }
}

class Flip implements Move {
  const Flip();
  
  void apply(SixStair s) {
    s.flip();
  }
  
  String toString() {
    return 'F';
  }
}

File: move.go
package sixstair

import (
	"strconv"
)

type Move int

func AllMoves() [11]Move {
	var res [11]Move
	for i := 1; i < 6; i++ {
		res[(i-1)*2] = Move(i)
	}
	res[10] = Move(0)
	return res
}

func (m Move) Apply(s *State) {
	if m == 0 {
		s.Flip()
	} else if m > 0 {
		for i := 0; i < int(m); i++ {
			s.Turn(true)
		}
	} else {
		for i := 0; i < -int(m); i++ {
			s.Turn(false)
		}
	}
}

func (m Move) String() string {
	if m == 0 {
		return "F"
	} else {
		return "T" + strconv.Itoa(int(m))
	}
}
